044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,342 @@
+1. we call `describeCluster` and get a response that says broker 1 and 2 are active
+2. broker 2 shuts down cleanly and is removed from the active set
+3. we call `describeLogDirs( brokerIds = [1,2] )` and only receive descriptions for logdirs on broker 1
+
+We propose introducing a configurable **throttle factor fallback** to be applied in situations where we don't have enough
+information to act. With a default value of 1.0 to optimistically allow all the quota to be used. This would allow
+users to opt in to more pessimistic behaviour like using a factor of 0.0 to prevent writes when we are in an unknown
+state. (note: we will also use the [Throttle Factor Validity Duration](#throttle-factor-validity-duration) to control
+when the fallback is applied)
+
+Throttle factor fallback can be in the range (0.0, 1.0)
+
+Example configuration: `client.quota.callback.static.throttle.factor.fallback=0.0`
+
+##### Throttle Factor Validity Duration
+
+To make the throttle factor logic more resilient in the face of transient failures we want to continue
+using a previously calculated valid throttle factor for some duration, rather than immediately using the 
+throttle factor fallback after a single failure.
+
+So we will enable the user to specify how long a valid throttle factor can be applied for with
+`client.quota.callback.static.throttle.factor.validity.duration`. Where the value is an ISO8601 duration.
+Default value "PT5M", 5 minutes.
+
+For example if validity duration is 2 minutes then we could expect this behaviour:
+
+1. throttle factor recalculation at minute 0 succeeded, calculated throttle factor 0.6 and use it
+2. throttle factor recalculation at minute 1 failed, continue using throttle factor 0.6
+3. throttle factor recalculation at minute 2 failed, continue using throttle factor 0.6
+4. throttle factor recalculation at minute 3 failed, use throttle factor fallback of 1.0
+5. throttle factor recalculation at minute 4 succeeded, calculated throttle factor 0.6 and use it
+
+## Configuration Summary
+
+|                                                                | type   | default | valid values      |                                                                                                                                  |
+|----------------------------------------------------------------|--------|---------|-------------------|----------------------------------------------------------------------------------------------------------------------------------|
+| client.quota.callback.static.storage.volume.source             | string | local   | local, cluster    | set to cluster to obtain volume descriptions from cluster, see [VolumeSource](#volume-source)                                    |
+| client.quota.callback.static.storage.check-interval            | long   | 0       | (0, ...)          | 0 means disabled, otherwise this is the milliseconds between polling to describe the cluster (this is an existing configuration) |
```

Comment:
Should the unit be in the name? E.g. `client.quota.callback.static.storage.check-interval-ms`

Reply:
As part of changing the proposal to discard the old implementation we changed this to `client.quota.callback.static.storage.check.interval` and changed the type to an ISO8601 duration string with a 1 minute default.